### ISSUE TITLE
Custom validator registration

### ISSUE THREAD

**Issue by SylvainCorlay** [ipython/traitlets#298]
I think that it would be interesting to push upstream the abiility that was added in jupyter-incubator's traittypes: registration of custom validators (non-cross) with a chaining `valid` method.

```python
from traitlets import HasTraits, TraitError
from traittypes import Array

def shape(*dimensions):
    def validator(trait, value):
        if value.shape != dimensions:
            raise TraitError('Expected an of shape %s and got and array with shape %s' % (dimensions, value.shape))
        else:
            return value
    return validator

class Foo(HasTraits):
    bar = Array(np.identity(2)).valid(shape(2, 2))
foo = Foo()

foo.bar = [1, 2]  # Should raise a TraitError
```

**Comment by SylvainCorlay** [comment]
cc @minrk

**Comment by minrk** [comment]
Looks nice to me, though I would use `.validate` instead of `.valid`, since `.valid` is not a verb.

If this works, doesn't that mean that we should be able to do:

```python
@bar.validate # or @bar.default
def _foo():
```

?

I use numpy a bunch in ipyparallel, but I've never used it with traits.

**Comment by SylvainCorlay** [comment]
- The problem is that `validate`  is already a public function in `TraitType` .
- Indeed I can use the custom cross validation with `@validate` but I was in a situation in bqplot where I needed to have

```
x = Array(one-dimensional)
y = Array(one or two dimensional, squeeze dimensions of length 1)
matrix = Array(shape(2, 2))
```

these custom validators are not cross-validators in that they don't require the other attributes of the HasTraits class and tend to be declaratively  specified for each attribute.

Example here:

https://github.com/bloomberg/bqplot/pull/305/files

**Comment by SylvainCorlay** [comment]
@minrk regarding the serialization, you are probably still using custom serialization routines to json + binary buffers, which could potentially be shared with ipywidgets.

I am still unsure if this will work with numpy datetime objects.

**Comment by SylvainCorlay** [comment]
If we introduce it in `5.0` then traittypes will need to remove their overload of `validate` calling the custom validator since the logic will likely be in `TraitType._validate`... This is an argument for including the logic now in traitlets, unless we don't pick the name `valid`.

What do you think of this @minrk?

**Comment by minrk** [comment]
No, we can't break usage of the current `@validate` name, we would need to use something else. Using a noun doesn't feel right to me, but I guess it's okay.

**Comment by SylvainCorlay** [comment]
Ah that is not what I meant.

**Comment by SylvainCorlay** [comment]
I mean't that `traittype.Array.validate` (the overloaded public method) uses the custom validators registered with `valid`. 

If we eventually have a similar API in base `TraitType`, registering things with `valid`, the validators list would be called in the private `_validate`, and if this list has the same name as in the one of Array, if will be called twice.

**Comment by minrk** [comment]
I see. We'd need to figure out a way to adopt the feature upstream safely, or have traittypes detect whether its additions are redundant, and make it conditional.

**Comment by SylvainCorlay** [comment]
Yeah, actually, if `Array.validators` become `TraitTypes._validators` if should be sufficient.

**Comment by rmorshea** [comment]
@SylvainCorlay why store the validators on TraitType instances, and not on the class?

Storing on TraitTypes seems a bit dangerous since a subclass calling `TraitType.valid` effects upstream classes, and other down stream ones that access that same trait.

**Comment by SylvainCorlay** [comment]
> @SylvainCorlay why store the validators on TraitType instances, and not on each HasTraits class?

> Storing on TraitTypes seems a bit dangerous since anything that calls TraitType.valid will effect all subclasses of the one class that happens to own the trait.

The custom validators don't have a reference on the TraitType instance. I don't think that it would make sense to make it a traittype instance thing.

**Comment by rmorshea** [comment]
@SylvainCorlay, I guess you could make the same argument against something like `TraitType.tag` so my criticism isn't valid. However, the issue I wanted to address with that comment was extendability of these chained validators in subclasses.

Without the ability to extend these validation chains people may be tempted to tamper with `TraitType._validators`. So my suggestion is that we implement something like this:

```python

class Parent(HasTraits):
	a = Array().valid(dtype(int)).tag(x=1)

class Child(Parent):
	a = Array(np.identity).valid(shape(2, 2), extend=True)
```

This is equally applicable to `tag`:

```python

class Parent(HasTraits):
	a = TraitType().tag(x=1)

class Child(Parent):
	a = TraitType().tag(y=2, extend=True)
```

The only difficulty being `tag(**metadata)`.

**Comment by rmorshea** [comment]
Additionally, the name `valid` is very confusing.

I think `has_valid` would make it significantly more distinct.
